src/Data/GADT/Coerce.hs

```diff
@@ -0,0 +1,107 @@
+{-# LANGUAGE CPP #-}
+{-# LANGUAGE GADTs #-}
+{-# LANGUAGE PolyKinds           #-}
+#if __GLASGOW_HASKELL__ >= 810
+{-# LANGUAGE StandaloneKindSignatures #-}
+#endif
+module Data.GADT.Coerce (
+    GCoercible (..),
+    defaultGcoercible,
+    ) where
+
+import Data.Functor.Product (Product (..))
+import Data.Functor.Sum     (Sum (..))
+import Data.IORef           (IORef)
+import Data.STRef           (STRef)
+import Data.Type.Equality   ((:~:) (..))
+import Data.Type.Coercion
+
+import Unsafe.Coerce        (unsafeCoerce)
+
+import Data.GADT.Compare    (GEq, geq)
+
+#if MIN_VERSION_base(4,10,0)
+import           Data.Type.Equality ((:~~:) (..))
+import qualified Type.Reflection    as TR
+#endif
+
+#if __GLASGOW_HASKELL__ >= 810
+import Data.Kind (Type, Constraint)
+#endif
+
+-- |A class for type-contexts which contain enough information
+-- to (at least in some cases) decide the coercibility of types
+-- occurring within them.
+#if __GLASGOW_HASKELL__ >= 810
+type GCoercible :: (k -> Type) -> Constraint
+#endif
+class GCoercible f where
+    gcoercible :: f a -> f b -> Maybe (Coercion a b)
+
+-- |If 'f' has a 'GEq' instance, this function makes a suitable default
+-- implementation of 'gcoercible'.
+defaultGcoercible :: GEq f => f a -> f b -> Maybe (Coercion a b)
+defaultGcoercible x y = fmap repr $ geq x y
+
+instance GCoercible ((:~:) a) where
+    gcoercible = defaultGcoercible
+
+#if MIN_VERSION_base(4,10,0)
+instance GCoercible ((:~~:) a) where
+    gcoercible = defaultGcoercible
+
+instance GCoercible TR.TypeRep where
+    gcoercible = defaultGcoercible
+#endif
+
+instance (GCoercible a, GCoercible b) => GCoercible (Sum a b) where
+    gcoercible (InL x) (InL y) = gcoercible x y
+    gcoercible (InR x) (InR y) = gcoercible x y
+    gcoercible _ _ = Nothing
+
+instance (GCoercible a, GCoercible b) => GCoercible (Product a b) where
+    gcoercible (Pair x y) (Pair x' y') = do
+        Coercion <- gcoercible x x'
+        Coercion <- gcoercible y y'
+        return Coercion
+
+instance GCoercible IORef where
+    gcoercible x y =
```

Comment:
I'll be more welcome to add this and `STRef` instance if the implementation for this function were in `base`, and I could use *Safe* (or at least *Trustworthy*) interface. (This package takes Safe Haskell seriously for time being).

Turns out there is for heterogenous equality for `StableName` (eqStableName), so figuring out coercion based on `STRef` equality kind of belongs to `base` as well.

As maintainer I'll be happy if burden of tracking GHC internals is in GHC-provided lib.

Reply:
I just had the same thought. The unsafe function should be where the implementation is, as those must be kept in sync?